### ISSUE TITLE
[CI] WatchBackwardsCompatibilityIT failures

### ISSUE THREAD

**Issue by martijnvg** [elastic/elasticsearch#40841]
Several bwc tests fail, because of connection timeouts. In the es logs of the upgraded node the following fatal error can be found:

```
[2019-04-03T22:02:01,777][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [upgraded-node-0] fatal error in thread [elasticsearch[upgraded-node-0][write][T#4]], exi
ting
java.lang.AssertionError: setIfMatch [3], currentDocTem [1]
        at org.elasticsearch.action.index.IndexRequest.writeTo(IndexRequest.java:657) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.action.DocWriteRequest.writeDocumentRequest(DocWriteRequest.java:244) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.action.bulk.BulkItemRequest.writeTo(BulkItemRequest.java:110) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.action.bulk.BulkShardRequest.writeTo(BulkShardRequest.java:73) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.action.support.replication.TransportReplicationAction$ConcreteShardRequest.writeTo(TransportReplicationAction.java:1276) ~[elasticsearch-6
.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.action.support.replication.TransportReplicationAction$ConcreteReplicaRequest.writeTo(TransportReplicationAction.java:1332) ~[elasticsearch
-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.OutboundMessage.writeMessage(OutboundMessage.java:70) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.OutboundMessage.serialize(OutboundMessage.java:53) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.OutboundHandler$MessageSerializer.get(OutboundHandler.java:107) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.OutboundHandler$MessageSerializer.get(OutboundHandler.java:93) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.OutboundHandler$SendContext.get(OutboundHandler.java:140) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.OutboundHandler.internalSendMessage(OutboundHandler.java:78) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.OutboundHandler.sendMessage(OutboundHandler.java:70) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.TcpTransport.sendRequestToChannel(TcpTransport.java:690) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.TcpTransport.sendRequestToChannel(TcpTransport.java:679) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.TcpTransport.access$300(TcpTransport.java:100) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.TcpTransport$NodeChannels.sendRequest(TcpTransport.java:272) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.transport.TransportService.sendRequestInternal(TransportService.java:626) ~[elasticsearch-6.7.1-SNAPSHOT.jar:6.7.1-SNAPSHOT]
        at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor.sendWithUser(SecurityServerTransportInterceptor.java:155) ~[?:?]
        at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor.access$400(SecurityServerTransportInterceptor.java:52) ~[?:?]
        at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$1.lambda$sendRequest$2(SecurityServerTransportInterceptor.java:129) ~[?:?]
        at org.elasticsearch.xpack.core.security.SecurityContext.executeAfterRewritingAuthentication(SecurityContext.java:111) ~[?:?]
```

I suspect that this assertion error is caused by watcher trying to update a watch status, because the watcher tests was the first one to fail with connection timeout. Also Watcher's [execution service](https://github.com/elastic/elasticsearch/blob/6.7/x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/execution/ExecutionService.java#L353) does not take into account nodes before 6.6.0 that don't support `if_seq_no` and `if_primary_term` parameters on write requests. 

build url: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.7+bwc-tests/226/console

**Comment by martijnvg** [comment]
This should be fixed by #40888

**Comment by davidkyle** [comment]
I have seen a similar error in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.7+bwc-tests/364/console

```
[2019-05-08T23:06:37,431][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [upgraded-node-0] fatal error in thread [elasticsearch[upgraded-node-0][write][T#1]], exiting
java.lang.AssertionError: setIfMatch [630], currentDocTem [1]
	at org.elasticsearch.action.index.IndexRequest.writeTo(IndexRequest.java:657) ~[elasticsearch-6.7.3-SNAPSHOT.jar:6.7.3-SNAPSHOT]
	at org.elasticsearch.action.DocWriteRequest.writeDocumentRequest(DocWriteRequest.java:244) ~[elasticsearch-6.7.3-SNAPSHOT.jar:6.7.3-SNAPSHOT]
	at org.elasticsearch.action.bulk.BulkItemRequest.writeTo(BulkItemRequest.java:110) ~[elasticsearch-6.7.3-SNAPSHOT.jar:6.7.3-SNAPSHOT]
	at org.elasticsearch.action.bulk.BulkShardRequest.writeTo(BulkShardRequest.java:73) ~[elasticsearch-6.7.3-SNAPSHOT.jar:6.7.3-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ConcreteShardRequest.writeTo(TransportReplicationAction.java:1276) ~[elasticsearch-6.7.3-SNAPSHOT.jar:6.7.3-SNAPSHOT]
```

The error is on the one third upgraded node upgrading from 6.1.0 to 6.7.3, I grabbed relevant logs 

[rolling-upgrade-logs.zip](https://github.com/elastic/elasticsearch/files/3162664/rolling-upgrade-logs.zip)

**Comment by davidkyle** [comment]
I suspect https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.7+bwc-tests/366/console and https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.8+bwc-tests/27/console are the same root problem. In both cases a node seems to have died unexpectedly

**Comment by droberts195** [comment]
All 6.7 and 6.8 BWC builds still in the history (~11 days) have failed with this problem:

* https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.7+bwc-tests/
* https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.8+bwc-tests/

**Comment by henningandersen** [comment]
One more here from today:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.8+bwc-tests/62/

**Comment by jakelandis** [comment]
I can reproduce this locally with 
```
./gradlew :x-pack:qa:rolling-upgrade:with-system-key:v6.0.1#bwcTest -Dtests.class=org.elasticsearch.upgrades.WatcherUpgradeIT
```
but not 
```
./gradlew :x-pack:qa:rolling-upgrade:with-system-key:v6.6.0#bwcTest -Dtests.class=org.elasticsearch.upgrades.WatcherUpgradeIT
```

I am pretty confident that this error:
```
Standard error:
kill: (812909): No such process
```
is not related to anything with Watcher.  Still looking, but this could be an indication of more general rolling upgrade from early version of 6.x issues.

**Comment by matriv** [comment]
Another failure: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.8+bwc-tests/72/console
Full logs can be found [here](https://github.com/elastic/elasticsearch/files/3217603/bwc_failure.txt.zip)

**Comment by jakelandis** [comment]
Got to the root cause of this failure and logged a new issue: https://github.com/elastic/elasticsearch/issues/42561

This isn't really a Watcher issue, it shows up in the Watcher tests since Watcher is continually issuing partial document updates (to .watches) while in a mixed cluster. Real world reproduction results in an IllegalStateException but since the tests are run with the assertion, the assertion crashes the running JVM. 

This only impacts the 6.x branches and I am bit hesitant to simply mute all Watcher BWC tests (which would be required) since it discovered a real issue. We _could_ simply remove (or comment out) the [assertion](https://github.com/elastic/elasticsearch/blob/v6.8.0/server/src/main/java/org/elasticsearch/action/index/IndexRequest.java#L657) and I tests pass just fine without the assertion (though there are extra errors in the logs). 

I would prefer that we comment out the assertion with a link to the issue. This would allow the tests to pass without having to disable all of Watcher's BWC tests for reasons Watcher does not control. If there are other BWC tests that get caught up on this, this would allow them to pass too. 

Thoughts on commenting out the assertion in 6.8 ? (as opposed to muting the test)

**Comment by dnhatn** [comment]
Great find. Thanks @jakelandis. This is fixed by #42596.
